modules/alloydb/main.tf

```diff
@@ -0,0 +1,371 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  prefix = var.prefix == null ? "" : "${var.prefix}-"
+  # has_replicas = try(length(var.replicas) > 0, false)
+  is_regional = var.availability_type == "REGIONAL" ? true : false
+  # secondary instance type is aligned with cluster type unless apply is targeting a promotion, in that
+  # case cluster will be 'primary' while instance still 'secondary'.
+  secondary_instance_type = try(var.cross_region_replication.promote_secondary && google_alloydb_cluster.secondary[0].cluster_type == "SECONDARY" ? "SECONDARY" : google_alloydb_cluster.secondary[0].cluster_type, null)
+
+  users = {
+    for k, v in coalesce(var.users, {}) :
+    k => {
+      name     = k
+      password = try(v.type, "ALLOYDB_BUILT_IN") == "ALLOYDB_BUILT_IN" ? try(random_password.passwords[k].result, v.password) : null
+      roles    = v.roles
+      type     = coalesce(v.type, "ALLOYDB_BUILT_IN")
+    }
+  }
+}
+
+resource "google_alloydb_cluster" "primary" {
+  project          = var.project_id
+  cluster_id       = "${local.prefix}${var.cluster_name}"
+  database_version = var.database_version
+  deletion_policy  = var.deletion_policy
+  labels           = var.labels
+  location         = var.location
+
+  network_config {
+    network            = var.network_config.network
+    allocated_ip_range = var.network_config.allocated_ip_range
+  }
+
+  dynamic "automated_backup_policy" {
+    for_each = var.automated_backup_configuration.enabled ? { 1 = 1 } : {}
+    content {
+      enabled       = true
+      location      = try(var.automated_backup_configuration.location, var.location)
+      backup_window = var.automated_backup_configuration.backup_window
+      labels        = var.labels
+
+      dynamic "encryption_config" {
+        for_each = var.encryption_config != null ? { 1 = 1 } : {}
+        content {
+          kms_key_name = var.encryption_config.primary_kms_key_name
+        }
+      }
+
+      weekly_schedule {
+        days_of_week = var.automated_backup_configuration.weekly_schedule.days_of_week
+        start_times {
+          hours   = var.automated_backup_configuration.weekly_schedule.start_times.hours
+          minutes = var.automated_backup_configuration.weekly_schedule.start_times.minutes
+          seconds = var.automated_backup_configuration.weekly_schedule.start_times.seconds
+          nanos   = var.automated_backup_configuration.weekly_schedule.start_times.nanos
+        }
+      }
+
+      dynamic "quantity_based_retention" {
+        for_each = var.automated_backup_configuration.retention_count != null ? { 1 = 1 } : {}
```

Comment:
use `[""]` and `[]` over `{1=1}` and `{}`

Reply:
Aligned on all the dynamic blocks!